components/tidb_query/src/codec/mysql/time/mod.rs

```diff
@@ -824,6 +815,32 @@ impl ConvertTo<Decimal> for Time {
     }
 }
 
+impl ConvertTo<Json> for Time {
+    #[inline]
+    fn convert(&self, _: &mut EvalContext) -> Result<Json> {
+        let s = if self.time_type == TimeType::DateTime || self.time_type == TimeType::Timestamp {
+            // TODO: avoid this clone
+            let mut val = self.clone();
+            val.fsp = mysql::MAX_FSP as u8;
+            val.to_string()
+        } else {
+            self.to_string()
+        };
+        Ok(Json::String(s))
+    }
+}
+
+impl ConvertTo<MyDuration> for Time {
+    fn convert(&self, _: &mut EvalContext) -> Result<MyDuration> {
+        if self.is_zero() {
+            return Ok(MyDuration::zero());
+        }
+        let nanos = i64::from(self.time.num_seconds_from_midnight()) * NANOS_PER_SEC
+            + i64::from(self.time.nanosecond());
```

Comment:
Is `self.time.nanosecond` same as `frac := t.Time.Microsecond() * 1000`

Reply:
The `self.time.nanosecond` is not same as `t.Time.Microsecond() * 1000`, but we will get the correct value because of `MyDuration::from_nanos` == `MyDuration::from_micros(self.time.nanosecond/1000)`.